dartagnan/src/main/java/com/dat3m/dartagnan/encoding/EncodeSets.java

```diff
@@ -6,21 +6,17 @@
 import com.dat3m.dartagnan.wmm.Definition.Visitor;
 import com.dat3m.dartagnan.wmm.analysis.RelationAnalysis;
 import com.dat3m.dartagnan.wmm.utils.Tuple;
-import com.dat3m.dartagnan.wmm.utils.TupleSet;
-
-import java.util.ArrayList;
-import java.util.Collection;
-import java.util.HashMap;
-import java.util.HashSet;
-import java.util.List;
-import java.util.Map;
+
+import java.util.*;
+import java.util.function.Function;
 import java.util.stream.Stream;
 
 // Collections of relationships in a verification task that need to be constrained in an SMT-based verification.
 final class EncodeSets implements Visitor<Map<Relation, Stream<Tuple>>> {
 
     private final RelationAnalysis ra;
     List<Tuple> news;
+    Map<Relation, Function<Event, Collection<Tuple>>> outSetMap = new HashMap<>();
```

Comment:
Those functions indeed encapsulate closures with an eagerly-constructed `Map<Event,Collection<Event>>`.  I did not consider doing the collection lazily, because the full may set would have to be fully iterated on every update.
- At some point, I chose to outline expressions like `mayOutMap.getOrDefault(e, List.of())`, since a `null` return value does not serve any additional semantics.
- Although `getMayOut()` is currently implemented non-virtually, I consider to let `Knowledge` be extended for base relations.  When this happens, the `Function` interface will describe the structure of the result more precisely than `Map`.  In those cases, the function could be lazy (i.e. `id`,`po`).
- My choice came with the loss of `mayOutMap.keySet()` and `mayOutMap.values()`.  An additional method will be required to compute the domain and co-domain of the may set of a relation.

Maybe I should inject TODOs to tell code reviewers.

Reply:
> I did not consider doing the collection lazily, because the full may set would have to be fully iterated on every update.

The whole intention of the caching mechanism is to avoid expensive recomputations, currently exactly by not being lazy but eager.

> Although getMayOut() is currently implemented non-virtually, I consider to let Knowledge be extended for base relations. When this happens, the Function interface will describe the structure of the result more precisely than Map. In those cases, the function could be lazy (i.e. id,po).

I agree that there is potential for laziness in some cases (like base relations) and this is also what we eventually need to do for large benchmarks where memory becomes a serious issue (especially computing `ext` explicitly sucks).
But this will also affect the whole set of edges itself, not just the in/out sets.

To handle this properly, we will likely need our own graph implementations (similar to `TupleSet` but just as an interface with many different actual implementations from fully eager to fully lazy).
One big problem is that due to backward propagations, even may and must sets may take unnatural forms (e.g. `maySet(id)` may not contain all identities anymore).
We will likely need to devise a semi-lazy structure like `Sets.difference(lazyIdBaseMaySet, eagerDisabledSet)` to handle this efficiently (`id` is not problem, but `ext` would be).

I guess one could model all the laziness by reusing the guava library for `Sets` and `Function` for in/out sets, but somehow I feel that this won't work in the long run.

Anyway, we can keep this for now, but I would put at least a comment that says that the actual Functions (or their results) are eagerly computed and that is why we cache them for performance reasons!

Reply:
That being said, the construction of the function objects will be expensive most of the time.  Thus they should be cached for the duration of the lifetime of `EncodeSets`.